src/v/cloud_topics/objects/level_one_object.h

```diff
@@ -0,0 +1,392 @@
+    // size_t object_size = ...;
+    // auto iobuf = read_object(
+    //   handle,
+    //   {.offset = object_size - 1_KiB, .size = 1_KiB},
+    // );
+    // auto result = l1::object_index::read_footer(iobuf.share());
+    // if (std::holds_alternative<l1::object_index>(result)) {
+    //   return std::get<l1::object_index>(result);
+    // }
+    // size_t extra = std::get<size_t>(result);
+    // auto missing = read_object(
+    //   handle,
+    //   {.offset = object_size - 1_KiB - extra, .size = extra},
+    // );
+    // missing.append(std::move(iobuf));
+    // result = l1::object_index::read_footer(std::move(missing));
+    // return std::get<l1::object_index>(result);
+    // ```
+    static std::variant<object_index, size_t> read_footer(iobuf);
+};
+
+// A builder of an l1 object, which is a collection of partitions.
+//
+// An L1 object is a sequence of partition segments, each segment is a sequence
+// of batches within a partition. There is a special marker that is used to
+// seperate partitions and contain the footer/index information.
+//
+// NOTE: If there is an error at any point when using the builder, it must be
+// discarded. It's possible there is partially flushed data that would result
+// in an invalid file if resumed.
+//
+// NOTE: It's valid to call start_partition() with the same NTP multiple times
+// but the data *must* contain disjoint offset ranges. There is currently no
+// restriction that segments for the same NTP must be written in order.
+class object_builder {
+public:
+    object_builder() = default;
+    object_builder(const object_builder&) = default;
+    object_builder(object_builder&&) = delete;
+    object_builder& operator=(const object_builder&) = default;
+    object_builder& operator=(object_builder&&) = delete;
+    virtual ~object_builder() = default;
+
+    // Options for the object_builder
+    struct options {
+        constexpr static size_t default_indexing_frequency = 4_MiB;
+        // The frequency at which to index the object, in bytes.
+        size_t indexing_frequency = default_indexing_frequency;
+    };
+
+    // Create a new object_builder that writes to the given output stream.
+    //
+    // The returned object_builder must be closed before destructing.
+    static std::unique_ptr<object_builder>
+      create(ss::output_stream<char>, options);
+
+    // Start writing batches for this partition.
+    //
+    // This must be called before any add_batch() calls, and calling this
+    // after calling start_partition() implicitly ends the current partition
+    // and starts a new one.
+    virtual ss::future<> start_partition(model::ntp ntp) = 0;
+
+    // Append a kafka batch to the object. The batch here is expected to be:
+    //
+    //  - A raft data batch, (meaning the header's type is set to raft_data).
+    //  - The offsets in this batch are > than the previous batch in this
+    //    partition.
+    virtual ss::future<> add_batch(model::record_batch) = 0;
```

Comment:
we may need the capability to write multiple ntps in parallel? Otherwise the API forces the per ntp buffering requirement on to the callers IIUC.

Reply:
My thinking here is that we can have a method to concat builders, it's not to hard to update the footer, since that is the only places with pointers into the data. I've not added that here, because I'd like to see the need and how to structure the API before creating that method. It's not clear to me how we want to do the combining (in memory, on disk, from cloud directly, etc)